contracts/serverless-v2/CommonChainJobs.sol

```diff
@@ -232,141 +223,199 @@ contract CommonChainJobs is
+        bytes memory _signature,
+        uint256 _jobId,
+        bytes memory _output,
+        uint256 _totalTime,
+        uint8 _errorCode
+    ) external {
+        _submitOutput(_signature, _jobId, _output, _totalTime, _errorCode);
+    }
+
+    function isJobExecutor(
+        uint256 _jobId,
+        address _executor
+    ) public view returns (bool) {
+        return _isJobExecutor(_jobId, _executor);
+    }
+
+    //-------------------------------- external functions end ----------------------------------//
+
     //-------------------------------- Job end --------------------------------//
 
+
     //-------------------------------- Timeout start --------------------------------//
 
     event SlashedOnExecutionTimeout(
         uint256 indexed jobId,
-        uint256 indexed reqChainId,
         address[] executors
     );
 
     event GatewayReassigned(
         uint256 indexed jobId,
-        uint256 indexed reqChainId,
         address prevGatewayKey,
         address reporterGateway,
         uint8 sequenceId
     );
 
-    function slashOnExecutionTimeout(
-        uint256 _jobId,
-        uint256 _reqChainId
-    ) external {
-        uint256 key = getKey(_jobId, _reqChainId);
-        require(jobs[key].jobId > 0, "INVALID_JOB");
+    error InvalidJob();
+    error DeadlineNotOver();
+
+    //-------------------------------- internal functions start ----------------------------------//
+
+    // TODO: active jobs cannot be updated if deadlineover and 2 of the executor nodes haven't submitted response
+    function _slashOnExecutionTimeout(
+        uint256 _jobId
+    ) internal {
+        if(jobs[_jobId].jobId == 0)
+            revert InvalidJob();
 
         // check for time
-        require(
-            block.timestamp > jobs[key].execStartTime + jobs[key].deadline + EXECUTION_BUFFER_TIME,
-            "DEADLINE_NOT_OVER"
-        );
+        if(block.timestamp <= jobs[_jobId].execStartTime + jobs[_jobId].deadline + EXECUTION_BUFFER_TIME)
+            revert DeadlineNotOver();
 
-        delete jobs[key];
+        delete jobs[_jobId];
 
         // slash Execution node
 
-        uint256 len = selectedExecutors[key].length;
+        uint256 len = selectedExecutors[_jobId].length;
         for (uint256 index = 0; index < len; index++) {
-            address executorKey = selectedExecutors[key][index];
-            require(!hasExecutedJob[key][executorKey], "JOB_ALREADY_EXECUTED");
-            delete hasExecutedJob[key][executorKey];
-            executors.updateOnExecutionTimeoutSlash(executorKey);
+            address executorKey = selectedExecutors[_jobId][index];
+            executors.updateOnExecutionTimeoutSlash(executorKey, hasExecutedJob[_jobId][executorKey]);
+            delete hasExecutedJob[_jobId][executorKey];
         }
 
-        emit SlashedOnExecutionTimeout(_jobId, _reqChainId, selectedExecutors[key]);
+        emit SlashedOnExecutionTimeout(_jobId, selectedExecutors[_jobId]);
```

Comment:
I think, only those executors should be return in event those who have not submitted the output.

Reply:
now individual events will be emtted for all the executors to be slashed